opentelemetry-sdk/src/metrics/internal/histogram.rs

```diff
@@ -65,62 +60,34 @@ impl<T: Number> Buckets<T> {
             ..Default::default()
         }
     }
-
-    fn sum(&mut self, value: T) {
-        self.total += value;
-    }
-
-    fn bin(&mut self, idx: usize, value: T) {
-        self.counts[idx] += 1;
-        self.count += 1;
-        if value < self.min {
-            self.min = value;
-        }
-        if value > self.max {
-            self.max = value
-        }
-    }
 }
 
 /// Summarizes a set of measurements as a histogram with explicitly defined
 /// buckets.
 pub(crate) struct Histogram<T: Number> {
-    value_map: ValueMap<HistogramTracker<T>>,
+    value_map: ValueMap<Mutex<Buckets<T>>>,
     bounds: Vec<f64>,
     record_min_max: bool,
     record_sum: bool,
     start: Mutex<SystemTime>,
 }
 
 impl<T: Number> Histogram<T> {
-    pub(crate) fn new(boundaries: Vec<f64>, record_min_max: bool, record_sum: bool) -> Self {
-        // TODO fix the bug, by first removing NaN and only then getting buckets_count
-        // once we know the reason for performance degradation
-        let buckets_count = boundaries.len() + 1;
-        let mut histogram = Histogram {
+    pub(crate) fn new(mut bounds: Vec<f64>, record_min_max: bool, record_sum: bool) -> Self {
+        bounds.retain(|v| !v.is_nan());
```

Comment:
There are only 3 ways bounds can come from:
1. Default.  - In this case, sdk provides it so we ensure its already sorted and validated.
2. HistogramCreation API - In this case, user provides it, but while creating the instrument, sdk *should* validate it, and ensure its sorted/validated. If the validation fails, sdk can return the defaults or noop instrument. Opened this issue to track it: https://github.com/open-telemetry/opentelemetry-rust/issues/2286
3. Views - In this case also, user provides it, but we should do same as 2.

I believe this method is the wrong place to validate/filter bounds as it should be taken care elsewhere.

Not a blocker for this PR, just added a comment to make sure have same understanding.

Reply:
> I believe this method is the wrong place to validate/filter bounds as it should be taken care elsewhere.

💯